flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/legacy/metrics/MetricStore.java

```diff
@@ -156,6 +179,27 @@ public synchronized ComponentMetricStore getSubtaskMetricStore(
         return ComponentMetricStore.unmodifiable(task.getSubtaskMetricStore(subtaskIndex));
     }
 
+    public synchronized ComponentMetricStore getSubtaskAttemptMetricStore(
+            String jobID, String taskID, int subtaskIndex, int attemptId) {
+        JobMetricStore job = jobID == null ? null : jobs.get(jobID);
+        if (job == null) {
+            return null;
+        }
+        TaskMetricStore task = job.getTaskMetricStore(taskID);
+        if (task == null) {
+            return null;
+        }
+        SubtaskMetricStore subtask = task.getSubtaskMetricStore(subtaskIndex);
+        if (attemptId < 0) {
```

Comment:
A negative attempt number is considered valid before when there were constructors using -1 as the default value, and was used to acquire the subtask level metrics instead of the attempt level one. Now that we have removed those constructors, a negative attempt  number is indeed invalid. I'll cleanup the relating codes.